src/Api/HL/RSQLInput.php

```diff
@@ -78,186 +78,266 @@ public function getOperators(): array
+                ],
+                [
+                    'operator' => '=isnull=',
+                    'description' => 'is null',
+                    'sql_where_callable' => fn ($a, $b) => [
+                        [$DB::quoteName($a) => null]
+                    ],
                 ],
-            ],
-            [
-                'operator' => '=notnull=',
-                'description' => 'is not null',
-                'sql_where_callable' => fn ($a, $b) => [
-                    ['NOT' => [$DB::quoteName($a) => null]]
+                [
+                    'operator' => '=notnull=',
+                    'description' => 'is not null',
+                    'sql_where_callable' => fn ($a, $b) => [
+                        ['NOT' => [$DB::quoteName($a) => null]]
+                    ],
                 ],
-            ],
-            [
-                // Empty string or null
-                'operator' => '=empty=',
-                'description' => 'is empty',
-                'sql_where_callable' => fn ($a, $b) => [
-                    [
-                        'OR' => [
-                            [$DB::quoteName($a) => ''],
-                            [$DB::quoteName($a) => null],
+                [
+                    // Empty string or null
+                    'operator' => '=empty=',
+                    'description' => 'is empty',
+                    'sql_where_callable' => fn ($a, $b) => [
+                        [
+                            'OR' => [
+                                [$DB::quoteName($a) => ''],
+                                [$DB::quoteName($a) => null],
+                            ]
                         ]
-                    ]
+                    ],
                 ],
-            ],
-            [
-                'operator' => '=notempty=',
-                'description' => 'is not empty',
-                'sql_where_callable' => fn ($a, $b) => [
-                    [
-                        'AND' => [
-                            [$DB::quoteName($a) => ['<>', '']],
-                            [$DB::quoteName($a) => ['<>', null]],
+                [
+                    'operator' => '=notempty=',
+                    'description' => 'is not empty',
+                    'sql_where_callable' => fn ($a, $b) => [
+                        [
+                            'AND' => [
+                                [$DB::quoteName($a) => ['<>', '']],
+                                [$DB::quoteName($a) => ['<>', null]],
+                            ]
                         ]
-                    ]
+                    ],
                 ],
-            ],
-        ];
+            ];
+        }
+        return $operators;
     }
 
-    /**
-     * @return array
-     * @phpstan-return {field: string, value: string, callable: callable}[]
-     */
-    private function getFilters()
+    public function getGroupedFilters(string $filter_string, array $flattened_props)
```

Comment:
It is pretty hard to read. Could you add a test on this method, with complex cases?